core/src/main/java/cz/o2/proxima/transform/RenameTransformation.java

```diff
@@ -0,0 +1,98 @@
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package cz.o2.proxima.transform;
+
+import cz.o2.proxima.annotations.Evolving;
+import cz.o2.proxima.functional.BiFunction;
+import cz.o2.proxima.functional.UnaryFunction;
+import cz.o2.proxima.repository.AttributeDescriptor;
+import cz.o2.proxima.repository.Repository;
+import cz.o2.proxima.storage.StreamElement;
+import java.util.Map;
+import lombok.extern.slf4j.Slf4j;
+
+/** A stateless transformation that performs transformation from one attribute to another. */
+@Evolving
+@Slf4j
+public class RenameTransformation implements ElementWiseTransformation {
+
+  private final UnaryFunction<AttributeDescriptor<?>, AttributeDescriptor<?>> descTransform;
+  private final BiFunction<String, AttributeDescriptor<?>, String> nameTransform;
+
+  public RenameTransformation(
+      UnaryFunction<AttributeDescriptor<?>, AttributeDescriptor<?>> descTransform,
+      BiFunction<String, AttributeDescriptor<?>, String> nameTransform) {
+
+    this.descTransform = descTransform;
+    this.nameTransform = nameTransform;
+  }
+
+  @Override
+  public void setup(Repository repo, Map<String, Object> cfg) {
+    // nop
+  }
+
+  @Override
+  public int apply(StreamElement input, Collector<StreamElement> collector) {
+    @SuppressWarnings("unchecked")
+    AttributeDescriptor<Object> desc =
+        (AttributeDescriptor<Object>) descTransform.apply(input.getAttributeDescriptor());
+
+    if (input.isDelete()) {
+      collector.collect(
+          input.isDeleteWildcard()
+              ? StreamElement.deleteWildcard(
+                  input.getEntityDescriptor(),
+                  desc,
+                  input.getUuid(),
+                  input.getKey(),
+                  nameTransform.apply(input.getAttribute(), input.getAttributeDescriptor()),
+                  input.getStamp())
+              : StreamElement.delete(
+                  input.getEntityDescriptor(),
+                  desc,
+                  input.getUuid(),
+                  input.getKey(),
+                  nameTransform.apply(input.getAttribute(), input.getAttributeDescriptor()),
+                  input.getStamp()));
+    } else {
+      collector.collect(
+          StreamElement.upsert(
+              input.getEntityDescriptor(),
+              desc,
+              input.getUuid(),
+              input.getKey(),
+              nameTransform.apply(input.getAttribute(), input.getAttributeDescriptor()),
+              input.getStamp(),
+              remapPayload(input, desc)));
+    }
+    return 1;
```

Comment:
Notice: this return code is still not clear for me :(

Reply:
This is due to the fact, that the rename can be asynchronous, and therefore the collector might not get invoked prior to returning from this method. The return value signals how many elements to wait for.